app/routes.py

```diff
@@ -2,35 +2,119 @@
 from app import crypto
 from app import email
 from datetime import datetime
-from sqlalchemy import desc
+from sqlalchemy import desc, and_, or_
+from sqlalchemy.orm import contains_eager
 
 import json
 import os
 import math
+import inspect
+
+
+import logging
+logging.basicConfig()
+# logging.getLogger('sqlalchemy.engine').setLevel(logging.INFO)
 
 @app.route('/')
 @app.route('/index')
 @login_required
 def index():
-    user = User.query.filter_by(username=current_user.username).first()
+    user = User.query.\
+        filter_by(username=current_user.username).\
+        join(Hospital).\
+        options(contains_eager(User.hospital)).\
+        first()
+
     if user.is_admin:
         return render_template('admin_index.html', title='Home')
     elif not user.is_verified:
         return render_template("404.html")
-    else:
-        user = User.query.filter_by(username=current_user.username).first()
-        user_hospital = Hospital.query.filter_by(id=user.hospital_id).first()
-        
-        hospital = {
-            "hospital_name": user_hospital.name
-        }
-        return render_template('index.html', title='Home', hospital=hospital)
+
+    ppe_types = PPE.query.\
+        outerjoin(Wants, and_(Wants.hospital_id == user.hospital.id, Wants.ppe_id == PPE.id)).\
+        outerjoin(Has, and_(Has.hospital_id == user.hospital.id, Has.ppe_id == PPE.id)).\
+        options(contains_eager(PPE.wants)).\
+        options(contains_eager(PPE.has)).\
+        all()
+
+        # filter(or_(Exchange.hospital1==user.hospital.id, Exchange.hospital2 == user.hospital.id)).\
+    exchange_subquery = Exchange.query.\
+        join(PPE).\
+        join(Hospital, or_(Exchange.hospital1 == Hospital.id, Exchange.hospital2 == Hospital.id)).\
+        options(contains_eager(Exchange.ppe_ref)).\
+        options(contains_eager(Exchange.hospital1_ref)).\
+        options(contains_eager(Exchange.hospital2_ref)).\
+        subquery()
+    
+    exchanges = Exchanges.query.\
+        filter(Exchanges.status.in_([EXCHANGE_COMPLETE, EXCHANGE_COMPLETE_ADMIN, EXCHANGE_IN_PROGRESS, EXCHANGE_UNVERIFIED])).\
+        join(exchange_subquery, Exchange.exchange_id == Exchanges.id).\
+        options(contains_eager(Exchanges.exchange)).\
+        order_by(Exchanges.id.asc()).\
+        all()
+
+    items = { 'actionable': [], 'pending': [], 'complete': [] }
+    for ex in exchanges:
```

Comment:
This actionability algorithm is pretty janky and far too complicated. However, I'm pretty sure it's correct, and it was the simplest solution that didn't require rewriting the DB schemea.